store/src/main/java/org/apache/rocketmq/store/queue/BatchConsumeQueue.java

```diff
@@ -1005,4 +1005,83 @@ public void cleanSwappedMap(long forceCleanSwapIntervalMs) {
     public MappedFileQueue getMappedFileQueue() {
         return mappedFileQueue;
     }
+
+    @Override
+    public long estimateMessageCount(long from, long to, MessageFilter filter) {
+        long physicalOffsetFrom = from * CQ_STORE_UNIT_SIZE;
+        long physicalOffsetTo = to * CQ_STORE_UNIT_SIZE;
+        List<MappedFile> mappedFiles = mappedFileQueue.range(physicalOffsetFrom, physicalOffsetTo);
+        if (mappedFiles.isEmpty()) {
+            return -1;
+        }
+
+        boolean sample = false;
+        long match = 0;
+        long raw = 0;
+
+        for (MappedFile mappedFile : mappedFiles) {
+            int start = 0;
+            int len = mappedFile.getFileSize();
+            // First file segment
+            if (mappedFile.getFileFromOffset() <= physicalOffsetFrom) {
+                start = (int) (physicalOffsetFrom - mappedFile.getFileFromOffset());
+                if (mappedFile.getFileFromOffset() + mappedFile.getFileSize() >= physicalOffsetTo) {
+                    // Current mapped file covers search range completely.
+                    len = (int) (physicalOffsetTo - physicalOffsetFrom);
+                } else {
+                    len = mappedFile.getFileSize() - start;
+                }
+            }
+
+            // Scan partial of the last file segment
+            if (0 == start && mappedFile.getFileFromOffset() + mappedFile.getFileSize() > physicalOffsetTo) {
+                len = (int) (physicalOffsetTo - mappedFile.getFileFromOffset());
+            }
+
+            SelectMappedBufferResult slice = mappedFile.selectMappedBuffer(start, len);
+            if (null != slice) {
+                try {
+                    ByteBuffer buffer = slice.getByteBuffer();
+                    int current = 0;
+                    while (current < len) {
+                        // Skip physicalOffset and message length fields.
+                        buffer.position(current + 8 + 4);
+                        long tagCode = buffer.getLong();
```

Comment:
Since there may be multiple messages in one BCQ index, IMO, It would be better to consider using the batch size field to estimate a more accurate value.